### ISSUE TITLE
Do not rely on `Content-Length` header for datasource response size metric

### ISSUE THREAD

**Issue by dafydd-t** [grafana/grafana#33372]
**What would you like to be added**:

Currently, the `datasourceResponseSummary` metric, observed in [`datasource_cache.go`](https://github.com/grafana/grafana/blob/master/pkg/models/datasource_cache.go#L101) is populated with the value parsed from the `Content-Length`  header of the response. 

We have seen a small number of examples in Grafana Cloud where some datasources to not send this header with the response, and we are therefore unable to see the size of the responses. 

One alternative is to measure the length of the payload ourselves, similar to [this RPC handler in weaveworks/common](https://github.com/weaveworks/common/blob/bd288de53d57de300fa286688ce2fc935687213f/middleware/grpc_stats.go#L51).

**Why is this needed**:

We would like to make the metrics for our Grafana Cloud instances more robust. By scanning the response to calculate its size, we would not rely on the existence of this header to get an accurate value for this metric.

**Comment by marefr** [comment]
The link to weaveworks/common seems to refer to  gRPC. Was that intended? Have an idea of how to solve this with http?

**Comment by dafydd-t** [comment]
Yes, I am aware that the the example uses gRPC, it was just an example of a pattern.

I've dug through `http.Response` a little more, and we might be seeing this behaviour for a different reason. First:

- It looks like the response *does not* rely on the Content-Length header anyway. `http.Response.ContentLength` is computed by the package and overrides any `Content-Length` header provided in the response anyway ([link](https://golang.org/src/net/http/response.go#L46)). So we're getting this desired behaviour from the `http` package.


So why is the value sometimes 0 when we know there is content? We may be seeing this behaviour because of the following `Response.Uncompressed` bool. The doc for this property states that `ContentLength` is set to -1 if this property is true. ([link to doc](https://golang.org/src/net/http/response.go#L89)) This `Uncrompressed` property is true if the `http` package decompressed the response before returning it (this seems to happen when the `http` package transparently asks for a compressed response).

We could stop this behaviour by enabling `http.Transport.DisableCompression` ([link to doc](https://golang.org/src/net/http/transport.go#L181)) which would force `http.Transport` not to transparently accept gzipped responses. This could of course have negative impact on datasource response size and time (plus I don't know how difficult it would be to plumb this option in)

If we were to try and continue with this, I think we should:
1. Verify that this `Uncompressed` boolean is the reason why we don't see a content length
2. Figure out how much work it would be to plumb this remedy in - if we need to make changes to external libraries  it may not be worth it. But it looks like we control the `Transport` config [around here](https://github.com/grafana/grafana/blob/10a46063153e1a6938f2c6a8f394b4a7aad9fec7/pkg/models/datasource_cache.go#L168).

**Comment by marefr** [comment]
We disable gzip for data source proxy request, see https://github.com/grafana/grafana/blob/a40aef0822c27fa9d145e8410f3b7979feb8446b/pkg/middleware/gziper.go#L13. What requests are we talking about, the ones going through /api/ds/query or /api/tsdb/query?

**Comment by bergquist** [comment]
https://github.com/felixge/httpsnoop solves this fairly well. The weaveworks server implements it here https://github.com/weaveworks/common/blob/bd288de53d57de300fa286688ce2fc935687213f/middleware/instrument.go#L67-L72

I'm not sure we need a histogram for each route in Grafana (considering the amount of routes) but it would be good to have one histogram per datasources and one for all _none datasource_ HTTP requests and also include the response size in the request logger so we could find requests that return "too much" data.

**Comment by marefr** [comment]
@bergquist https://github.com/felixge/httpsnoop if for incoming request/response. We need it for outgoing requests/responses, right? 

Possibly we could implement a `ReadCloser` that counts the number of bytes read that the metrics middleware could use for getting the actual response size/bytes read. Otherwise we always need to read the full response in the middleware and that doesn't feel great from a performance/memory consumption perspective.

**Comment by marefr** [comment]
The existing metrics Middleware is here: https://github.com/grafana/grafana/blob/main/pkg/infra/httpclient/httpclientprovider/datasource_metrics_middleware.go

**Comment by bergquist** [comment]
> @bergquist https://github.com/felixge/httpsnoop if for incoming request/response. We need it for outgoing requests/responses, right?

Primarily for outgoing requests but incoming would also be interesting. But that could be another issue
